File: web/lib/sheets.ts
// Server-side client for the Google Apps Script Web App that backs our sheet.
// Never import from the browser — the SECRET lives in env and we relay
// requests through Next API routes.

import type { BookingPayload, BookingResult, Slot, AdminBooking } from "./types";

const WEBHOOK_URL = process.env.SHEETS_WEBHOOK_URL;
const SECRET = process.env.SHEETS_WEBHOOK_SECRET;

function assertConfigured() {
  if (!WEBHOOK_URL || !SECRET) {
    throw new Error(
      "Sheets backend not configured. Set SHEETS_WEBHOOK_URL and SHEETS_WEBHOOK_SECRET in .env.local"
    );
  }
}

export async function fetchSlots(params: {
  sport: string;
  date: string;
}): Promise<Slot[]> {
  assertConfigured();
  const url = `${WEBHOOK_URL}?action=slots&sport=${encodeURIComponent(params.sport)}&date=${encodeURIComponent(params.date)}`;
  const res = await fetch(url, {
    method: "GET",
    cache: "no-store",
  });
  if (!res.ok) throw new Error(`Sheets fetch failed: ${res.status}`);
  const json = await res.json();
  return (json.slots as Slot[]) || [];
}

export async function createBooking(
  payload: BookingPayload
): Promise<BookingResult> {
  assertConfigured();
  // Apps Script doPost can't read custom headers reliably, so we send the
  // shared secret in the body. The Web App is HTTPS-only so this is fine.
  const res = await fetch(WEBHOOK_URL!, {
    method: "POST",
    headers: { "Content-Type": "application/json" },
    body: JSON.stringify({ action: "book", _secret: SECRET, ...payload }),
  });
  const json = (await res.json()) as BookingResult & {
    booking_id?: string;
  };
  if (!res.ok || !json.success) {
    return { success: false, error: json.error || "Booking failed" };
  }
  return { success: true, bookingId: json.bookingId || json.booking_id };
}

export async function fetchAllBookings(): Promise<AdminBooking[]> {
  assertConfigured();
  const url = `${WEBHOOK_URL}?action=admin_bookings&_secret=${encodeURIComponent(SECRET!)}`;
  const res = await fetch(url, { method: "GET", cache: "no-store" });
  if (!res.ok) throw new Error(`Admin fetch failed: ${res.status}`);
  const json = await res.json();
  return (json.bookings as AdminBooking[]) || [];
}

File: whatsapp-bot/bot/sheets.py
"""Python client for the Apps Script webhook — mirrors web/lib/sheets.ts."""
from __future__ import annotations

import httpx

from .config import settings


async def fetch_slots(sport: str, date: str) -> list[dict]:
    if not settings.SHEETS_WEBHOOK_URL:
        return []
    url = f"{settings.SHEETS_WEBHOOK_URL}?action=slots&sport={sport}&date={date}"
    async with httpx.AsyncClient(timeout=10.0) as client:
        r = await client.get(url)
        r.raise_for_status()
        data = r.json()
    return data.get("slots", []) or []


async def create_booking(payload: dict) -> dict:
    if not settings.SHEETS_WEBHOOK_URL or not settings.SHEETS_WEBHOOK_SECRET:
        return {"success": False, "error": "Sheets backend not configured"}
    body = {"action": "book", "_secret": settings.SHEETS_WEBHOOK_SECRET, **payload}
    async with httpx.AsyncClient(timeout=15.0) as client:
        r = await client.post(settings.SHEETS_WEBHOOK_URL, json=body)
    return r.json()


async def fetch_bookings_today() -> list[dict]:
    """For reminders. Add a 'bookings_today' action to Code.gs to support this."""
    if not settings.SHEETS_WEBHOOK_URL:
        return []
    url = f"{settings.SHEETS_WEBHOOK_URL}?action=bookings_today"
    try:
        async with httpx.AsyncClient(timeout=10.0) as client:
            r = await client.get(url)
            r.raise_for_status()
            data = r.json()
        return data.get("bookings", []) or []
    except Exception:
        return []
